numba/errors.py

```diff
@@ -188,6 +201,43 @@ class ConstantInferenceError(NumbaError):
     """
 
 
+class InternalError(NumbaError):
+    """
+    For wrapping internal error occured within the compiler
+    """
+    def __init__(self, exception):
+        super(InternalError, self).__init__(str(exception))
+        self.old_exception = exception
+
+
+def _format_msg(fmt, args, kwargs):
+    return fmt.format(*args, **kwargs)
+
+
+@contextlib.contextmanager
+def new_error_context(fmt_, *args, **kwargs):
+    """
+    A contextmanager that prepend contextual information to any exception
+    raised within.  If the exception type is not an instance of NumbaError,
+    it will be wrapped into a InternalError.   The exception class can be
+    changed by providing a "errcls_" keyword argument with the exception
```

Comment:
I'm letting usage like:

```python
with new_error_context('lowering "{inst}" at {loc}', inst=inst,
                                      loc=self.loc, errcls_=defaulterrcls):
````

so that the underscore-suffixed keywords are for special parameters and others are for the format-string.